File: scripts/transport-v4-ready.mjs
import assert from "node:assert/strict";
import { createPrivateKey, createPublicKey, sign, timingSafeEqual } from "node:crypto";
import { createRequire } from "node:module";
import { decodeMap, encodeCBOR, mapFromNames, strictHex } from "./transport-v4-codec.mjs";
import { evaluateDomain } from "./transport-v4-domains.mjs";
import { buildNoiseCorpus } from "./transport-v4-noise.mjs";
import { encodeRecordLayout } from "./transport-v4-records.mjs";

// Public fixture composition only. Supplied digests are not authenticated
// credentials. This helper cannot publish a Session or grant READY authority.
const require = createRequire(new URL("../flowersec-ts/package.json", import.meta.url));
const { ed25519 } = require("@noble/curves/ed25519.js");
const hex = bytes => Buffer.from(bytes).toString("hex");
const flip = value => { const changed = strictHex(value); changed[0] ^= 0x80; return hex(changed); };
const privatePrefix = Buffer.from("302e020100300506032b657004220420", "hex");

function strictProof(proof, message, key) {
  if (key.length !== 32 || proof.length !== 64) return false;
  try {
    for (const encoded of [key, proof.subarray(0, 32)]) {
      const point = ed25519.Point.fromBytes(encoded, false);
      if (point.is0() || !point.isTorsionFree() || !Buffer.from(point.toBytes()).equals(encoded)) return false;
    }
    return ed25519.verify(proof, message, key, { zip215: false });
  } catch { return false; }
}

export function verifyReadyPlan(schema) {
  const plan = schema.ready_vector_plan;
  assert.equal(plan.status, "public_fixture_ready_composition_only");
  assert.match(plan.source, /synthetic context digests/u);
  assert.deepEqual(plan.roles.map(r => r.role), [0, 1]);
  assert.deepEqual(plan.selected_features, [0, 3]);
  assert.deepEqual(plan.negative_mutations, ["each_ready_byte", "each_truncation", "noncanonical_map", "context_substitution", "invalid_proof_valid_mac", "reflection"]);
  assert.equal(new Set(plan.roles.map(r => r.public_key_hex)).size, 2);
  for (const role of plan.roles) for (const key of ["seed_hex", "public_key_hex", "certificate_digest_hex"]) assert.equal(strictHex(role[key]).length, 32);
  assert.deepEqual(Object.keys(plan.context).sort(), ["admission_binding_hex", "fsa_digest_hex", "fsb_digest_hex"]);
  for (const value of Object.values(plan.context)) assert.equal(strictHex(value).length, 32);
  const proof = schema.frame_maps.ReadyProofInput, mac = schema.frame_maps.ReadyMACInput;
  for (const [id, field] of Object.entries(proof.fields)) assert.deepEqual(mac.fields[id], field);
  assert.deepEqual(proof.required, [0,1,2,3,4,5,6,7]);
  assert.deepEqual(mac.required, [0,1,2,3,4,5,6,7,8,9]);
  assert.equal(schema.frame_maps.READY.fields[0].length, 64);
  assert.equal(schema.frame_maps.READY.fields[1].length, 32);
}

export function readyMap(schema, name, context) {
  const values = Object.fromEntries(Object.values(schema.frame_maps[name].fields).map(field => [field.name,
    field.type === "bytes" ? { $bytes: context[field.name + "_hex"] } : context[field.name],
  ]));
  const bytes = encodeCBOR(mapFromNames(schema, name, values));
  decodeMap(schema, name, bytes);
  return bytes;
}

export function readyMaterial(schema, context, proof) {
  const proofInput = readyMap(schema, "ReadyProofInput", context);
  const message = evaluateDomain(schema, "ready_identity", { proof: proofInput });
  const key = evaluateDomain(schema, "ready_key", {
    profile: context.crypto_profile_id, role: context.role,
    handshake_hash: strictHex(context.handshake_hash_hex), context_digest: strictHex(context.transport_context_digest_hex),
    epoch_root: strictHex(context.epoch_root_hex),
  });
  const out = { proof_input_hex: hex(proofInput), signature_message_hex: message.input_hex, key_info_hex: key.input_hex, key_hex: key.output_hex };
  if (proof !== undefined) {
    const macInput = readyMap(schema, "ReadyMACInput", { ...context, identity_proof_hex: hex(proof) });
    const mac = evaluateDomain(schema, "ready_mac", { mac_input: macInput, ready_key: strictHex(key.output_hex) });
    Object.assign(out, { mac_input_hex: hex(macInput), mac_message_hex: mac.input_hex, confirmation_mac_hex: mac.output_hex });
  }
  return out;
}

export function verifyReadyReference(schema, context, bytes) {
  try {
    const received = decodeMap(schema, "READY", bytes), fields = schema.frame_maps.READY.fields;
    const named = Object.fromEntries(Object.entries(fields).map(([id, field]) => [field.name, received.get(BigInt(id))]));
    const material = readyMaterial(schema, context, named.identity_proof);
    return timingSafeEqual(named.confirmation_mac, strictHex(material.confirmation_mac_hex))
      && strictProof(named.identity_proof, strictHex(material.signature_message_hex), strictHex(context.public_key_hex));
  } catch { return false; }
}

export function buildReadyCorpus(schema) {
  verifyReadyPlan(schema);
  const plan = schema.ready_vector_plan, noise = buildNoiseCorpus(schema), vectors = [], negatives = [];
  const encode = (proof, mac) => readyMap(schema, "READY", { identity_proof_hex: hex(proof), confirmation_mac_hex: hex(mac) });
  for (const transcript of noise.transcripts) for (const role of plan.roles) for (const features of plan.selected_features) {
    const context = { ...plan.context, role: role.role, crypto_profile_id: transcript.profile,
      handshake_hash_hex: transcript.handshake_hash_hex, transport_context_digest_hex: noise.inputs.context_digest_hex,
      certificate_digest_hex: role.certificate_digest_hex, selected_features: features,
      epoch_root_hex: transcript.initial_root_hex, public_key_hex: role.public_key_hex };
    const privateKey = createPrivateKey({ key: Buffer.concat([privatePrefix, strictHex(role.seed_hex)]), format: "der", type: "pkcs8" });
    assert.equal(hex(createPublicKey(privateKey).export({ format: "der", type: "spki" }).subarray(-32)), role.public_key_hex);
    const proof = sign(null, strictHex(readyMaterial(schema, context).signature_message_hex), privateKey);
    const material = readyMaterial(schema, context, proof), ready = encode(proof, strictHex(material.confirmation_mac_hex));
    assert.equal(ready.length, 103);
    assert.ok(verifyReadyReference(schema, context, ready));
    const envelope = encodeRecordLayout(schema.envelope.layout, { payload_length: ready.length, frame_type: schema.frame_types.READY });
    const id = `${transcript.id}_ready_role${role.role}_features${features}`;
    vectors.push({ id, noise_transcript: transcript.id, context, signing_seed_hex: role.seed_hex, ...material,
      identity_proof_hex: hex(proof), ready_hex: hex(ready), envelope_header_hex: hex(envelope), wire_hex: hex(Buffer.concat([envelope, ready])) });
    const add = (suffix, bytes, patch = {}, reason = suffix) => {
      assert.equal(verifyReadyReference(schema, { ...context, ...patch }, bytes), false, id + suffix);
      negatives.push({ id: `${id}_${suffix}`, source: id, ready_hex: hex(bytes), context_patch: patch, reason, expected_error: "ready_rejected" });
    };
    for (let i = 0; i < ready.length; i++) {
      const changed = Buffer.from(ready); changed[i] ^= 0x80;
      add(`byte_${i}`, changed, {}, "wire_mutation");
      add(`truncate_${i}`, ready.subarray(0, i), {}, "truncation");
    }
    const pairs = [...decodeMap(schema, "READY", ready)].map(([k,v]) => Buffer.concat([encodeCBOR(k), encodeCBOR(v)]));
    for (const [name, bytes] of [
      ["suffix", Buffer.concat([ready, Buffer.from([0])])],
      ["reordered", Buffer.concat([Buffer.from([0xa2]), ...pairs.toReversed()])],
      ["duplicate", Buffer.concat([Buffer.from([0xa3]), pairs[0], ...pairs])],
      ["unknown", Buffer.concat([Buffer.from([0xa3]), ...pairs, Buffer.from([2, 0])])],
      ["overlong_key", Buffer.concat([Buffer.from([0xa2, 0x18, 0]), ready.subarray(2)])],
      ["indefinite", Buffer.concat([Buffer.from([0xbf]), ready.subarray(1), Buffer.from([0xff])])],
    ]) add(name, bytes, {}, "noncanonical_map");
    for (const [name, original] of Object.entries(context)) {
      const value = name.endsWith("_hex") ? flip(original) : name === "crypto_profile_id"
        ? Object.keys(schema.crypto_profiles).find(p => p !== original) : name === "role" ? 1 - original : original ^ 1;
      const patch = { [name]: value };
      add(`context_${name}`, ready, patch, "context_substitution");
      // A root holder can construct a correct MAC, but cannot reuse the old
      // identity proof on a different signed context or certificate key.
      if (Object.values(schema.frame_maps.ReadyProofInput.fields).some(f => (f.type === "bytes" ? f.name + "_hex" : f.name) === name) || name === "public_key_hex") {
        const changed = readyMaterial(schema, { ...context, ...patch }, proof);
        add(`context_${name}_valid_mac`, encode(proof, strictHex(changed.confirmation_mac_hex)), patch, "invalid_proof_valid_mac");
      }
    }
    const badProof = Buffer.alloc(64); badProof[0] = 1;
    add("identity_R_valid_mac", encode(badProof, strictHex(readyMaterial(schema, context, badProof).confirmation_mac_hex)), {}, "invalid_proof_valid_mac");
    add("unknown_feature", ready, { selected_features: 4 }, "invalid_context");
    add("invalid_role", ready, { role: 2 }, "invalid_context");
    add("unknown_profile", ready, { crypto_profile_id: "unknown" }, "invalid_context");
  }
  for (const vector of vectors) {
    const other = vectors.find(v => v.context.crypto_profile_id === vector.context.crypto_profile_id && v.context.selected_features === vector.context.selected_features && v.context.role !== vector.context.role);
    assert.equal(verifyReadyReference(schema, vector.context, strictHex(other.ready_hex)), false);
    negatives.push({ id: vector.id + "_reflection", source: vector.id, ready_hex: other.ready_hex, context_patch: {}, reason: "reflection", expected_error: "ready_rejected" });
  }
  assert.equal(new Set([...vectors, ...negatives].map(v => v.id)).size, vectors.length + negatives.length);
  return { schema_revision: schema.schema_revision, design_sha256: schema.design_sha256, coverage: plan.status, vectors, negatives,
    unverified: ["Authenticated FSB/FSA/certificate/admission and complete Flowersec Noise prologue", "One-shot dual READY owner, publication and authorization/resource/deadline gates", "Four-SDK full Noise-to-rekey composition and independent cryptographic qualification", "Runtime secret handling, provider and pairwise qualification"] };
}

File: flowersec-rust/tests/transport_v4_ready.rs
//! Public fixtures only; no credential trust or one-shot dual-READY authority.
#[allow(dead_code)]
#[path = "../src/protocol_v4_registry_generated.rs"]
mod registry;
#[path = "../src/strict_ed25519_v4_reference.rs"]
mod strict;

use hkdf::Hkdf;
use ring::hmac;
use serde_json::{Value, json};
use sha2::Sha256;

fn unhex(value: &Value) -> Option<Vec<u8>> {
    let text = value.as_str()?;
    if !text.len().is_multiple_of(2) {
        return None;
    }
    text.as_bytes()
        .as_chunks::<2>()
        .0
        .iter()
        .map(|pair| u8::from_str_radix(std::str::from_utf8(pair).ok()?, 16).ok())
        .collect()
}

fn hex(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

fn head(major: u8, value: u64) -> Vec<u8> {
    if value < 24 {
        return vec![major << 5 | value as u8];
    }
    for (width, ai) in [(1, 24), (2, 25), (4, 26), (8, 27)] {
        if width == 8 || value < 1_u64 << (width * 8) {
            return [&[major << 5 | ai][..], &value.to_be_bytes()[8 - width..]].concat();
        }
    }
    unreachable!()
}

struct Reference {
    maps: Value,
    domains: Value,
    profiles: Value,
}
impl Reference {
    fn fields(&self, name: &str) -> Vec<(u64, &Value)> {
        let mut fields: Vec<_> = self.maps[name]["fields"]
            .as_object()
            .unwrap()
            .iter()
            .map(|(id, f)| (id.parse::<u64>().unwrap(), f))
            .collect();
        fields.sort_by_key(|(id, _)| *id);
        fields
    }
    fn encode(&self, name: &str, context: &Value) -> Option<Vec<u8>> {
        let fields = self.fields(name);
        let mut out = head(5, fields.len() as u64);
        for (id, f) in fields {
            out.extend(head(0, id));
            let name = f["name"].as_str()?;
            match f["type"].as_str()? {
                "bytes" => {
                    let value = unhex(&context[format!("{name}_hex")])?;
                    if value.len() as u64 != f["length"].as_u64()? {
                        return None;
                    }
                    out.extend(head(2, value.len() as u64));
                    out.extend(value);
                }
                "text" => {
                    let value = context[name].as_str()?;
                    self.profiles.get(value)?;
                    out.extend(head(3, value.len() as u64));
                    out.extend(value.as_bytes());
                }
                kind @ ("uint8" | "uint64") => {
                    let value = context[name].as_u64()?;
                    if kind == "uint8" && value > 255 {
                        return None;
                    }
                    if let Some(mask) = f.get("bitmask")
                        && value & !mask.as_u64()? != 0
                    {
                        return None;
                    }
                    if let Some(allowed) = f.get("enum")
                        && !allowed
                            .as_object()?
                            .values()
                            .any(|v| v.as_u64() == Some(value))
                    {
                        return None;
                    }
                    out.extend(head(0, value));
                }
                _ => return None,
            }
        }
        Some(out)
    }
    fn decode(&self, mut wire: &[u8]) -> Option<Value> {
        let fields = self.fields("READY");
        wire = wire.strip_prefix(head(5, fields.len() as u64).as_slice())?;
        let mut out = json!({});
        for (id, f) in fields {
            if f["type"] != "bytes" {
                return None;
            }
            let length = usize::try_from(f["length"].as_u64()?).ok()?;
            wire = wire.strip_prefix(head(0, id).as_slice())?;
            wire = wire.strip_prefix(head(2, length as u64).as_slice())?;
            let value = wire.get(..length)?;
            out[format!("{}_hex", f["name"].as_str()?)] = hex(value).into();
            wire = &wire[length..];
        }
        wire.is_empty().then_some(out)
    }
    fn domain(&self, name: &str, values: &[(&str, &[u8])], role: u64) -> Option<Vec<u8>> {
        let spec = self
            .domains
            .as_array()?
            .iter()
            .find(|d| d["name"] == name)?;
        let mut out = unhex(&spec["label_bytes"])?;
        for part in spec["input_schema"]["parts"].as_array()? {
            match part["encoding"].as_str()? {
                "u8" => {
                    if let Some(allowed) = part.get("enum")
                        && !allowed.as_array()?.contains(&json!(role))
                    {
                        return None;
                    }
                    out.push(u8::try_from(role).ok()?);
                }
                "lp-map" | "lp-ascii" | "lp-bytes" => {
                    let value = values
                        .iter()
                        .find(|(key, _)| Some(*key) == part["name"].as_str())?
                        .1;
                    if let Some(length) = part.get("length")
                        && value.len() as u64 != length.as_u64()?
                    {
                        return None;
                    }
                    out.extend(u32::try_from(value.len()).ok()?.to_be_bytes());
                    out.extend(value);
                }
                _ => return None,
            }
        }
        Some(out)
    }
    fn material(&self, context: &Value, proof: &[u8]) -> Option<Value> {
        let role = context["role"].as_u64()?;
        let proof_input = self.encode("ReadyProofInput", context)?;
        let message = self.domain("ready_identity", &[("proof", &proof_input)], role)?;
        let info = self.domain(
            "ready_key",
            &[
                ("profile", context["crypto_profile_id"].as_str()?.as_bytes()),
                ("handshake_hash", &unhex(&context["handshake_hash_hex"])?),
                (
                    "context_digest",
                    &unhex(&context["transport_context_digest_hex"])?,
                ),
            ],
            role,
        )?;
        let root = unhex(&context["epoch_root_hex"])?;
        if root.len() != 32 {
            return None;
        }
        let mut key = [0; 32];
        Hkdf::<Sha256>::from_prk(&root)
            .ok()?
            .expand(&info, &mut key)
            .ok()?;
        let mut values = context.clone();
        values["identity_proof_hex"] = hex(proof).into();
        let mac_input = self.encode("ReadyMACInput", &values)?;
        let mac_message = self.domain("ready_mac", &[("mac_input", &mac_input)], role)?;
        let mac = hmac::sign(&hmac::Key::new(hmac::HMAC_SHA256, &key), &mac_message);
        Some(
            json!({"proof_input_hex":hex(&proof_input),"signature_message_hex":hex(&message),
            "key_info_hex":hex(&info),"key_hex":hex(&key),"mac_input_hex":hex(&mac_input),
            "mac_message_hex":hex(&mac_message),"confirmation_mac_hex":hex(mac.as_ref())}),
        )
    }
    fn verify(&self, context: &Value, wire: &[u8]) -> bool {
        (|| {
            let received = self.decode(wire)?;
            let proof = unhex(&received["identity_proof_hex"])?;
            let material = self.material(context, &proof)?;
            let mac = unhex(&received["confirmation_mac_hex"])?;
            hmac::verify(
                &hmac::Key::new(hmac::HMAC_SHA256, &unhex(&material["key_hex"])?),
                &unhex(&material["mac_message_hex"])?,
                &mac,
            )
            .ok()?;
            Some(strict::verify(
                &proof,
                &unhex(&material["signature_message_hex"])?,
                &unhex(&context["public_key_hex"])?,
            ))
        })()
        .unwrap_or(false)
    }
}

#[test]
fn shared_ready_composition_and_rejection() {
    let corpus: Value =
        serde_json::from_str(include_str!("../../testdata/transport_v4/ready.json")).unwrap();
    let r = Reference {
        maps: serde_json::from_str(registry::READY_REGISTRY_JSON).unwrap(),
        domains: serde_json::from_str(registry::DOMAIN_REGISTRY_JSON).unwrap(),
        profiles: serde_json::from_str(registry::CRYPTO_PROFILES_JSON).unwrap(),
    };
    let record: Value = serde_json::from_str(registry::RECORD_REGISTRY_JSON).unwrap();
    assert_eq!(corpus["schema_sha256"], registry::SCHEMA_SHA256);
    let vectors = corpus["vectors"].as_array().unwrap();
    assert_eq!(vectors.len(), 8);
    for v in vectors {
        let proof = unhex(&v["identity_proof_hex"]).unwrap();
        let m = r.material(&v["context"], &proof).unwrap();
        for (key, value) in m.as_object().unwrap() {
            assert_eq!(value, &v[key], "{} {key}", v["id"]);
        }
        assert_eq!(
            strict::sign(
                &unhex(&m["signature_message_hex"]).unwrap(),
                &unhex(&v["signing_seed_hex"]).unwrap()
            )
            .unwrap(),
            proof
        );
        let ready=r.encode("READY",&json!({"identity_proof_hex":hex(&proof),"confirmation_mac_hex":m["confirmation_mac_hex"]})).unwrap();
        let values = json!({"payload_length":ready.len(),"frame_type":registry::FRAME_READY});
        let mut envelope = Vec::new();
        for f in record["envelope"]["layout"].as_array().unwrap() {
            let value = f
                .get("const")
                .unwrap_or(&values[f["name"].as_str().unwrap()])
                .as_u64()
                .unwrap();
            let width = match f["type"].as_str().unwrap() {
                "uint8" => 1,
                "uint16_be" => 2,
                "uint32_be" => 4,
                _ => panic!("envelope"),
            };
            assert!(value < 1_u64 << (width * 8));
            envelope.extend(&value.to_be_bytes()[8 - width..]);
        }
        assert_eq!(hex(&ready), v["ready_hex"]);
        assert_eq!(hex(&envelope), v["envelope_header_hex"]);
        assert_eq!(hex(&[envelope, ready.clone()].concat()), v["wire_hex"]);
        assert!(r.verify(&v["context"], &ready));
    }
    let negatives = corpus["negatives"].as_array().unwrap();
    assert!(!negatives.is_empty());
    for n in negatives {
        assert_eq!(n["expected_error"], "ready_rejected");
        let source = vectors.iter().find(|v| v["id"] == n["source"]).unwrap();
        let mut context = source["context"].clone();
        for (key, value) in n["context_patch"].as_object().unwrap() {
            context[key] = value.clone();
        }
        assert!(
            !r.verify(&context, &unhex(&n["ready_hex"]).unwrap()),
            "{}",
            n["id"]
        );
    }
}
